pkg/lib/authn/mfa/service.go

```diff
@@ -55,12 +92,23 @@ func (s *Service) CreateDeviceToken(userID string, token string) (*DeviceToken,
 }
 
 func (s *Service) VerifyDeviceToken(userID string, token string) error {
-	err := s.RateLimiter.TakeToken(AntiBruteForceDeviceTokenBucket(userID))
+	perUserPerIP, perIP, err := s.reserveRateLimit("VerifyDeviceToken",
+		s.Config.RateLimits.DeviceToken.PerUserPerIP,
+		s.Config.RateLimits.DeviceToken.PerIP,
+		userID,
+	)
+	defer s.RateLimiter.Cancel(perUserPerIP)
+	defer s.RateLimiter.Cancel(perIP)
+
 	if err != nil {
 		return err
 	}
 
 	_, err = s.DeviceTokens.Get(userID, token)
+	if errors.Is(err, ErrDeviceTokenNotFound) {
+		perUserPerIP.Consume()
+		perIP.Consume()
```

Comment:
Both perUserPerIP and perIP is nullable.

Reply:
If `reserveRateLimit` return nil `err`, both `perUserPerIP` and `perIP` is reserved successfully and non-nil.